### ISSUE TITLE
Please read & provide the following

### ISSUE THREAD

**Issue by WorstCodeWay** [facebookresearch/detectron2#3079]
If you do not know the root cause of the problem, please post according to this template:

## Instructions To Reproduce the Issue:

I trained mask-rcnn with my own dataset with pre-trained model *COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x.yaml*, and the result is good as expected when inferring with `DefaultPredictor`. But the result is not correct after I export trained model into TorchScript and infer in `libtorch`. My steps as following:

1. Train my model based on pre-trained model, the generated model is in
*tools/output/model_final.pth*
2. Convert model into TorchScript:
```bash
$./tools/deploy/export_model.py --config-file configs/COCO-InstanceSegmentation/mask_rcnn_R_50_FPN_3x.yaml     --output ./output --export-method tracing --format torchscript     MODEL.WEIGHTS tools/output/model_final.pth   MODEL.DEVICE cuda
```
There is a file *output/model.ts* which I think is generated.

3. Load this model in c++ codes:
**Note** I add `ResizeImage` function to make it same with `DefaultPredictor` behaviors.
```c++
#include <opencv2/opencv.hpp>
#include <torch/script.h> // One-stop header.
#include <c10/cuda/CUDAStream.h>
#include <torch/csrc/autograd/grad_mode.h>
#include <torch/csrc/jit/runtime/graph_executor.h>

// only needed for export_method=tracing
#include <torchvision/vision.h> // @oss-only

#include <iostream>
#include <memory>

void ResizeImage(cv::Mat &img, int minSide, int maxSide, cv::Size &size)
{
  int width = img.cols;
  int height = img.rows;
  int newWidth, newHeight;

  float scale = minSide * 1.0 / std::min(height, width);
  if(height < width)
  {
    newHeight = minSide;
    newWidth = scale * width;
  }
  else
  {
    newHeight = scale * height;
    newWidth = minSide;
  }
  if(std::max(newHeight, newWidth) > maxSide)
  {
    scale = maxSide * 1.0 / std::max(newHeight, newWidth);
    newHeight = newHeight * scale;
    newWidth = newWidth * scale;
  }
  newWidth = static_cast<int>(newWidth + 0.5);
  newHeight = static_cast<int>(newHeight + 0.5);

  size = cv::Size(newWidth, newHeight);
  cv::resize(img, img, size);
}

c10::IValue get_caffe2_tracing_inputs(cv::Mat& img, c10::Device device)
{
  const int height = img.rows;
  const int width = img.cols;
  // FPN models require divisibility of 32.
  // Tracing mode does padding inside the graph, but caffe2_tracing does not.
  assert(height % 32 == 0 && width % 32 == 0);
  const int channels = 3;

  auto input =
      torch::from_blob(img.data, {1, height, width, channels}, torch::kUInt8);
  // NHWC to NCHW
  input = input.to(device, torch::kFloat).permute({0, 3, 1, 2}).contiguous();

  std::array<float, 3> im_info_data{height * 1.0f, width * 1.0f, 1.0f};
  auto im_info =
      torch::from_blob(im_info_data.data(), {1, 3}).clone().to(device);
  return std::make_tuple(input, im_info);
}
c10::IValue get_tracing_inputs(cv::Mat& img, c10::Device device)
{
  const int height = img.rows;
  const int width = img.cols;
  const int channels = 3;

  std::cout << "width " << img.cols << " height " << img.rows << std::endl;
  cv::Size newSize;
  ResizeImage(img, 800, 1333, newSize);
  std::cout << "width " << img.cols << " height " << img.rows << std::endl;

  auto input =
      torch::from_blob(img.data, {newSize.height, newSize.width, channels}, torch::kUInt8);
  // HWC to CHW
  input = input.to(device, torch::kFloat).permute({2, 0, 1}).contiguous();
  return input;
}

int main(int argc, const char* argv[]) {
  if (argc < 3) {
    std::cerr << "usage: example-app <path-to-exported-script-module>\n";
    return -1;
  }
  std::string model_file = argv[1];
  std::string image_file = argv[2];
  std::string export_method = (argc == 4 ? argv[3] : "");

  bool is_caffe2 = (export_method == "caffe2_tracing");

  torch::jit::getBailoutDepth() = 1;
  torch::autograd::AutoGradMode guard(false);
  torch::jit::script::Module module;
  try 
  {
    // Deserialize the ScriptModule from a file using torch::jit::load().
    module = torch::jit::load(model_file);
    assert(module.buffers().size() > 0);
    // Assume that the entire model is on the same device.
    // We just put input to this device.
    auto device = (*std::begin(module.buffers())).device();

    cv::Mat imgSource = cv::imread(image_file, cv::IMREAD_COLOR);
    cv::Mat inputImg = imgSource;
    c10::IValue inputs = is_caffe2 ? get_caffe2_tracing_inputs(inputImg, device)
                          : get_tracing_inputs(inputImg, device);

    // run the network
    auto output = module.forward({inputs});
    if (device.is_cuda())
      c10::cuda::getCurrentCUDAStream().synchronize();
    
    auto outputs = output.toTuple()->elements();
    std::cout << "Number of output tensors: " << outputs.size() << std::endl;
    at::Tensor bbox, pred_classes, pred_masks, scores;
    // parse Mask R-CNN outputs

    // parse Mask R-CNN outputs
    if (is_caffe2)
    {
      bbox = outputs[0].toTensor(), scores = outputs[1].toTensor(),
      pred_classes = outputs[2].toTensor(), pred_masks = outputs[3].toTensor();
    }
    else
    {
      bbox = outputs[0].toTensor(), pred_classes = outputs[1].toTensor(),
      pred_masks = outputs[2].toTensor(), scores = outputs[3].toTensor();
      // outputs[-1] is image_size, others fields ordered by their field name in
      // Instances
    }

    std::cout << "bbox: " << bbox.toString() << " " << bbox.sizes() << std::endl;
    std::cout << "scores: " << scores.toString() << " " << scores.sizes() << std::endl;
    std::cout << "pred_classes: " << pred_classes.toString() << " "
              << pred_classes.sizes() << std::endl;
    std::cout << "pred_masks: " << pred_masks.toString() << " " << pred_masks.sizes()
              << std::endl;

    int num_instances = bbox.sizes()[0];
    std::cout << bbox << std::endl;

    // iterate by each dimension
    for(int num = 0; num < bbox.sizes()[0]; ++num)
    {
      float x1 = bbox[num][0].item().toFloat();
      float y1 = bbox[num][1].item().toFloat();
      float x2 = bbox[num][2].item().toFloat();
      float y2 = bbox[num][3].item().toFloat();

      cv::rectangle(imgSource, cv::Point2f(x1, y1), cv::Point2f(x2, y2), cv::Scalar(255));

      std::cout << num << " score: " << scores[num].item().toFloat() << std::endl;
    }
    cv::imshow("input", imgSource);
    cv::waitKey();
  }
  catch (const c10::Error& e)
  {
    std::cerr << e.what();
    return -1;
  }

  return 0;
}
```
After build this `.cpp` file into executable, executing it:
```bash
./test ../model.ts ../20210111150947_ori.png
```
3. __Full logs__ or other relevant observations:
```
width 1280 height 1024
width 1000 height 800
Number of output tensors: 5
bbox: CUDAFloatType [17, 4]
scores: CUDAFloatType [17]
pred_classes: CUDALongType [17]
pred_masks: CUDAFloatType [17, 1, 28, 28]
 173.0938  170.6849  276.8962  261.2823
 123.7460  139.4473  228.1219  229.3445
 271.4507  238.8707  399.7458  318.4761
 227.5246  210.3228  338.0922  285.1438
 418.8994  320.4454  553.0139  398.8542
 342.6711  252.0423  374.9548  321.4144
 667.2379  394.2019  798.9390  485.8072
  48.2644   97.1920  135.0796  179.2941
 350.5841  307.5519  439.0551  349.1892
 153.7327  143.1602  292.3083  210.5630
 606.2017  356.9136  749.5900  440.7903
 494.2209  268.0332  600.7437  384.0996
  82.3484  134.3959  184.6238  171.0598
  87.6828  125.3200  168.4965  192.0337
 845.4465  627.1760  863.3810  659.3284
  52.6549  105.3091  180.7573  201.0848
 344.9618  267.2150  446.8616  354.6699
[ CUDAFloatType{17,4} ]
0 score: 0.0809258
1 score: 0.0680464
2 score: 0.0645862
3 score: 0.0600019
4 score: 0.0591206
5 score: 0.0590583
6 score: 0.0578627
7 score: 0.056263
8 score: 0.0555598
9 score: 0.0542852
10 score: 0.053981
11 score: 0.0538849
12 score: 0.0531877
13 score: 0.0528937
14 score: 0.0515238
15 score: 0.0509743
16 score: 0.0504861
```

## Expected behavior:
**Predicting with DefaultPredictor in python**
![Screenshot from 2021-05-23 23-35-30](https://user-images.githubusercontent.com/17172534/119267021-a600a400-bc1f-11eb-9c1b-ab5007bdca52.png)

But get the wrong result with converted `TorchScript`:
![Screenshot from 2021-05-23 23-32-14](https://user-images.githubusercontent.com/17172534/119266890-283c9880-bc1f-11eb-97e4-2a97b9d608c1.png)

## Environment:

Paste the output of the following command:
```bash
wget -nc -nv https://github.com/facebookresearch/detectron2/raw/master/detectron2/utils/collect_env.py && python collect_env.py

wget -nc -nv https://github.com/facebookresearch/detectron2/raw/master/detectron2/utils/collect_env.py && python collect_env.py
2021-05-23 23:33:25 URL:https://raw.githubusercontent.com/facebookresearch/detectron2/master/detectron2/utils/collect_env.py [7340/7340] -> "collect_env.py" [1]
----------------------  ------------------------------------------------------------------------------------------------------
sys.platform            linux
Python                  3.6.9 (default, Jan 26 2021, 15:33:00) [GCC 8.4.0]
numpy                   1.19.5
detectron2              0.4 @/home/isee/Documents/deeplearning/detectron2/detectron2/detectron2
Compiler                GCC 7.5
CUDA compiler           not available
DETECTRON2_ENV_MODULE   <not set>
PyTorch                 1.8.1+cu102 @/home/isee/Documents/deeplearning/detectron2/venv/lib/python3.6/site-packages/torch
PyTorch debug build     False
GPU available           True
GPU 0                   GeForce RTX 2060 (arch=7.5)
CUDA_HOME               /usr/local/cuda
Pillow                  8.2.0
torchvision             0.9.1+cu102 @/home/isee/Documents/deeplearning/detectron2/venv/lib/python3.6/site-packages/torchvision
torchvision arch flags  3.5, 5.0, 6.0, 7.0, 7.5
fvcore                  0.1.5.post20210508
iopath                  0.1.8
cv2                     4.5.2
----------------------  ------------------------------------------------------------------------------------------------------
PyTorch built with:
  - GCC 7.3
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191122 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v1.7.0 (Git Hash 7aed236906b1f7a05c0917e5257a1af05e9ff683)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 10.2
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_70,code=sm_70
  - CuDNN 7.6.5
  - Magma 2.5.2
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=10.2, CUDNN_VERSION=7.6.5, CXX_COMPILER=/opt/rh/devtoolset-7/root/usr/bin/c++, CXX_FLAGS= -Wno-deprecated -fvisibility-inlines-hidden -DUSE_PTHREADPOOL -fopenmp -DNDEBUG -DUSE_KINETO -DUSE_FBGEMM -DUSE_QNNPACK -DUSE_PYTORCH_QNNPACK -DUSE_XNNPACK -O2 -fPIC -Wno-narrowing -Wall -Wextra -Werror=return-type -Wno-missing-field-initializers -Wno-type-limits -Wno-array-bounds -Wno-unknown-pragmas -Wno-sign-compare -Wno-unused-parameter -Wno-unused-variable -Wno-unused-function -Wno-unused-result -Wno-unused-local-typedefs -Wno-strict-overflow -Wno-strict-aliasing -Wno-error=deprecated-declarations -Wno-stringop-overflow -Wno-psabi -Wno-error=pedantic -Wno-error=redundant-decls -Wno-error=old-style-cast -fdiagnostics-color=always -faligned-new -Wno-unused-but-set-variable -Wno-maybe-uninitialized -fno-math-errno -fno-trapping-math -Werror=format -Wno-stringop-overflow, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.8.1, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=ON, USE_NNPACK=ON, USE_OPENMP=ON, 
```

If your issue looks like an installation issue / environment issue,
please first check common issues in https://detectron2.readthedocs.io/tutorials/install.html#common-installation-issues

**Comment by ppwwyyxx** [comment]
Since inputs are resized, you should draw predicted boxes on the resized image.

**Comment by WorstCodeWay** [comment]
> Since inputs are resized, you should draw predicted boxes on the resized image.

Not exactly:
I have resized input image and draw bounding box on that input image, but the boxes have wrong locations and sizes.

> `export_model.py --run-eval` can verify that the exported model produces same results. To report an issue we'll need runnable code (with necessary model, data, and hopefully in python) that verifies they produce different results.

By the way, I want to  try `--run-eval`, but how to do it in my own dataset? Or is there a way to predict on one test image?

**Comment by ppwwyyxx** [comment]
https://github.com/facebookresearch/detectron2/blob/42ef395311e8f4d9ff189416ba89f7c9c3d07258/tools/deploy/export_model.py#L226-L240

Here you can run `exported_model` and `torch_model` on same inputs and verify their outputs are close enough.
Input formats are documented in https://detectron2.readthedocs.io/en/latest/tutorials/models.html#model-input-format
